src/services/kratos/totp.ts

```diff
@@ -0,0 +1,89 @@
+import { UiNodeTextAttributes } from "@ory/client"
+import { LikelyNoUserWithThisPhoneExistError } from "@domain/auth/errors"
+
+import { kratosPublic } from "./private"
+import {
+  AuthenticationKratosError,
+  MissingTotpKratosError,
+  UnknownKratosError,
+} from "./errors"
+
+export const kratosInitiateTotp = async (token: SessionToken) => {
+  try {
+    const res = await kratosPublic.createNativeSettingsFlow({ xSessionToken: token })
+    const totpAttributes = res.data.ui.nodes.find(
+      (node) => (node.attributes as UiNodeTextAttributes).id === "totp_secret_key",
+    )
+    if (!totpAttributes) {
+      return new MissingTotpKratosError()
+    }
+
+    const totpSecret = (totpAttributes.attributes as UiNodeTextAttributes).text
+      .text as TotpSecret
+    return { totpSecret, flow: res.data.id as FlowId }
+  } catch (err) {
+    return new UnknownKratosError(err)
+  }
+}
+
+export const kratosValidateTotp = async ({
+  authToken,
+  totpCode,
+  flow,
+}: {
+  authToken: SessionToken
+  totpCode: string
+  flow: string
+}) => {
+  try {
+    await kratosPublic.updateSettingsFlow({
+      flow,
+      updateSettingsFlowBody: {
+        method: "totp",
+        totp_code: totpCode,
+      },
+      xSessionToken: authToken,
+    })
+  } catch (err) {
+    return new UnknownKratosError(err)
+  }
+}
```

Comment:
should we check the result of the response to be cautious if the kratos api changes? the happy path appears to return `resp.data.state === "success"`.

Reply:
good point, I'll make the changes